app/editor.js

```diff
@@ -16,36 +19,130 @@ ved.params = function() {
     }, {});
 };
 
-ved.select = function(spec) {
+ved.mode = function() {
+  var sel = ved.$d3.select('.sel_mode').node(),
+      idx = sel.selectedIndex,
+      newMode = sel.options[idx].value;
+
+  if (ved.currentMode === newMode)
+    return;
+  ved.currentMode = newMode;
+
+  if (ved.currentMode === 'vega') {
+    ved.vgEditor.setOptions({
+      readOnly: false,
+      highlightActiveLine: true,
+      highlightGutterLine: true
+    });
+    ved.$d3.select('.vega-editor').attr('class', 'vega-editor vega');
+    ved.$d3.select('.vg-spec .ace_content').attr('class', 'ace_content');
+  } else if (ved.currentMode === 'vega-lite') {
+    ved.vgEditor.setOptions({
+      readOnly: true,
+      highlightActiveLine: false,
+      highlightGutterLine: false
+    });
+
+    ved.$d3.select('.vg-spec .ace_content').attr('class', 'ace_content disabled');
+    ved.$d3.select('.vega-editor').attr('class', 'vega-editor vega-lite');
+
+    if (ved.vlEditor.getValue().length > 0) {
+      ved.parseVl();
+    } else {
+      ved.resetView();
+      ved.vgEditor.setValue('');
+    }
+  } else {
+    console.warn('Unknown mode ' + ved.currentMode);
+  }
+
+  ved.editorVisibility();
+};
+
+ved.switchToVega = function() {
+  var sel = ved.$d3.select('.sel_mode').node();
+  sel.selectedIndex = 0;
+  ved.mode();
+};
+
+// Changes visibility of vega editor in vl mode
+ved.editorVisibility = function() {
+  if (ved.vgHidden && ved.currentMode === 'vega-lite') {
+    ved.$d3.select('.vg-spec').style('display', 'none');
+    ved.$d3.select('.vl-spec')
+      .style('flex', '1 1 auto');
+  } else {
+    ved.$d3.select('.vg-spec').style('display', 'block');
+    ved.resizeVlEditor();
+  }
+  ved.resize();
+};
+
+ved.selectVg = function(spec) {
   var desc = ved.$d3.select('.spec_desc');
 
   if (spec) {
-    ved.editor.setValue(spec);
-    ved.editor.gotoLine(0);
+    ved.vgEditor.setValue(spec);
+    ved.vgEditor.gotoLine(0);
     desc.html('');
-    ved.parse();
+    ved.parseVg();
     return;
   }
-  
-  var sel = ved.$d3.select('.sel_spec').node(),
+
+  var sel = ved.$d3.select('.sel_vg_spec').node(),
       idx = sel.selectedIndex;
   spec = d3.select(sel.options[idx]).datum();
 
   if (idx > 0) {
-    d3.xhr(ved.uri(spec), function(error, response) {
-      ved.editor.setValue(response.responseText);
-      ved.editor.gotoLine(0);
-      ved.parse(function() { desc.html(spec.desc || ''); });
+    d3.xhr(ved.uriVg(spec), function(error, response) {
+      ved.vgEditor.setValue(response.responseText);
+      ved.vgEditor.gotoLine(0);
+      ved.parseVg(function() { desc.html(spec.desc || ''); });
     });
   } else {
-    ved.editor.setValue('');
-    ved.editor.gotoLine(0);
+    ved.vgEditor.setValue('');
+    ved.vgEditor.gotoLine(0);
+    ved.resetView();
+  }
+};
+
+ved.selectVl = function(spec) {
```

Comment:
`ved.selectVg` and `ved.selectVl` appear to be nearly identical. Can we extract common functionality to a separate parameterized method instead? If we were to modify how we use ace in the future, for example, this would reduce the number of places we'd have to make the change.